Config.txt
# Configuration file for GRSISpoon Calibration Sort.

# Lines begining # are ignored
# Items should be presented as KEY\nITEM
# List of all possible keyss and item formats  :

# SIM_CLOVER_AB_EFF
# SIM_CRYSTAL_EFF
# EXP_CLOVER_AB_EFF
# EXP_CRYSTAL_EFF

# Physics/DAQ Settings
INTEGRATION
500

DISPERSION
4

# Hit Detection
CHARGE_THRESH
100

ENERGY_THRESH
5.0

# Source Specification
# First two lines in list will be used for peak search, rest will be fitted afterwards
60CO
1173.237 1332.501

152EU FRONT
121.7817 1408.006 244.6975 344.2785 411.116 778.9040 964.079 1112.074 867.378 1299.14 

152EU BACK
344.2785 1408.006 244.6975 411.116 778.9040 964.079 1112.074 867.378 1299.14

133BA
276.398 356.017 80.9971 302.853 383.851

# Parameters for Calibration and fitting functions
MAX_TIME
36000.0

TIME_BINS
20

FIT_TEMP_SPECTRA
1

CHARGE_BINS
16384

CHARGE_BINS2D
1024

CHARGE_MAX
1500000

WAVE_CHARGE_MAX
4096

# Peak Search stuff
EN_SEARCH_THRESH
0.055  

EN_SEARCH_SIGMA
10

WAVE_SEARCH_THRESH
0.01 

WAVE_SEARCH_SIGMA
20

TIG_GAIN_EST
0.1603     

TIG_WAVE_GAIN_EST
0.6

ENERGY_SIGMA_ZERO
0.45

ENERGY_SIGMA_1MEV
0.45  

WAVE_SIGMA_ZERO
1.5     

WAVE_SIGMA_1MEV
0.4 

# Checks on fit quality

GAUS_HEIGHT_MIN
10    

GAUS_CSPD_MAX
50

GAUS_SIGMA_MIN
250

# minimum counts for fitting of histogram to be attempted
MIN_FIT_COUNTS
500

# Values for GeTiming Sort
GE_TIMING_GATE_CENTRE
121.8

GE_TIMING_GATE_WIDTH
5.0

# Seg Core 2D Calibration stuff
# SEG_CORE_FIT_ORDER should be 0, 1, or 2.    
SEG_CORE_FIT_ORDER
2

# reference efficiency values at 1332.5keV 

SIM_CLOVER_AB_EFF
1 0.5
2 0.5
3 0.5
4 0.5
5 0.5
6 0.5
7 0.5
8 0.5
9 0.5
10 0.5
11 0.5
12 0.5
13 0.5
14 0.5
15 0.5
16 0.5

EXP_CLOVER_AB_EFF
1 0.75
2 0.75
3 0.75
4 0.75
5 0.75
6 0.75
7 0.75
8 0.75
9 0.75
10 0.75
11 0.75
12 0.75
13 0.75
14 0.75
15 0.75
16 0.75

SIM_CRYSTAL_EFF
1 0 0.06
1 1 0.06
1 2 0.06
1 3 0.06
2 0 0.06
2 1 0.06
2 2 0.06
2 3 0.06
3 0 0.06
3 1 0.06
3 2 0.06
3 3 0.06
4 0 0.06
4 1 0.06
4 2 0.06
4 3 0.06
5 0 0.06
5 1 0.06
5 2 0.06
5 3 0.06
6 0 0.06
6 1 0.06
6 2 0.06
6 3 0.06
7 0 0.06
7 1 0.06
7 2 0.06
7 3 0.06
8 0 0.06
8 1 0.06
8 2 0.06
8 3 0.06
9 0 0.06
9 1 0.06
9 2 0.06
9 3 0.06
10 0 0.06
10 1 0.06
10 2 0.06
10 3 0.06
11 0 0.06
11 1 0.06
11 2 0.06
11 3 0.06
12 0 0.06
12 1 0.06
12 2 0.06
12 3 0.06
13 0 0.06
13 1 0.06
13 2 0.06
13 3 0.06
14 0 0.06
14 1 0.06
14 2 0.06
14 3 0.06
15 0 0.06
15 1 0.06
15 2 0.06
15 3 0.06
16 0 0.06
16 1 0.06
16 2 0.06
16 3 0.06

EXP_CRYSTAL_EFF
1 0 0.065
1 1 0.065
1 2 0.065
1 3 0.065
2 0 0.065
2 1 0.065
2 2 0.065
2 3 0.065
3 0 0.065
3 1 0.065
3 2 0.065
3 3 0.065
4 0 0.065
4 1 0.065
4 2 0.065
4 3 0.065
5 0 0.065
5 1 0.065
5 2 0.065
5 3 0.065
6 0 0.065
6 1 0.065
6 2 0.065
6 3 0.065
7 0 0.065
7 1 0.065
7 2 0.065
7 3 0.065
8 0 0.065
8 1 0.065
8 2 0.065
8 3 0.065
9 0 0.065
9 1 0.065
9 2 0.065
9 3 0.065
10 0 0.065
10 1 0.065
10 2 0.065
10 3 0.065
11 0 0.065
11 1 0.065
11 2 0.065
11 3 0.065
12 0 0.065
12 1 0.065
12 2 0.065
12 3 0.065
13 0 0.065
13 1 0.065
13 2 0.065
13 3 0.065
14 0 0.065
14 1 0.065
14 2 0.065
14 3 0.065
15 0 0.065
15 1 0.065
15 2 0.065
15 3 0.065
16 0 0.065
16 1 0.065
16 2 0.065
16 3 0.065

CRYSTAL_FHWM_1332
1 0 2.3
1 1 2.3
1 2 2.3
1 3 2.3
2 0 2.3
2 1 2.3
2 2 2.3
2 3 2.3
3 0 2.3
3 1 2.3
3 2 2.3
3 3 2.3
4 0 2.3
4 1 2.3
4 2 2.3
4 3 2.3
5 0 2.3
5 1 2.3
5 2 2.3
5 3 2.3
6 0 2.3
6 1 2.3
6 2 2.3
6 3 2.3
7 0 2.3
7 1 2.3
7 2 2.3
7 3 2.3
8 0 2.3
8 1 2.3
8 2 2.3
8 3 2.3
9 0 2.3
9 1 2.3
9 2 2.3
9 3 2.3
10 0 2.3
10 1 2.3
10 2 2.3
10 3 2.3
11 0 2.3
11 1 2.3
11 2 2.3
11 3 2.3
12 0 2.3
12 1 2.3
12 2 2.3
12 3 2.3
13 0 2.3
13 1 2.3
13 2 2.3
13 3 2.3
14 0 2.3
14 1 2.3
14 2 2.3
14 3 2.3
15 0 2.3
15 1 2.3
15 2 2.3
15 3 2.3
16 0 2.3
16 1 2.3
16 2 2.3
16 3 2.3







Options.h
// Header file for stuff related to run configuration and command line options
// ---------------------------------------------------------------------------

// --------------------------------------------------------
// Definitions:
// --------------------------------------------------------

// Printing info
#define PRINT_OUTPUT 1
#define PRINT_VERBOSE 0
#define PRINT_FREQ 50000
// Main loop control
#define MAX_EVENTS 0
#define DEBUG_TREE_LOOP 0
// ROOT Stuff
#define ROOT_VIRT_SIZE    1024u*1024u*1024u     // 1x10^7 or ~10Mb seems to run fast-ish but not freeze the system completely.
                                     // that's on my 6Gb 2.6GHz i5 (YMMV)
                                     // Correction: for a longer run file 10^9 was needed
// Stuff about the experimental setup
#define CLOVERS  16
#define CRYSTALS  4
#define SEGS      8
#define INTEGRATION 500  // Integration factor applied to charge values (K)
#define DISPERSION 4     // Dispersion applied to charge before histo incremented
// Constants
#define PI 3.14159265359

//Other stuff
#define CHAR_BUFFER_SIZE 1024

#define FOLD_MAX 128

// Item numbers for source records
#define SOURCE_60CO 0 
#define SOURCE_152EU_FRONT 1
#define SOURCE_152EU_BACK 2
#define SOURCE_133BA 3

// Information for calib and eff
// --------------------------------------------

// Plotting
#define PLOT_FITS 0             // plot fits, all chans plotted if below items = 0
#define PLOT_CALIB 0            // Plot calibration
#define PLOT_RESIDUAL 0
#define PLOT_CALIB_SUMMARY 1    // Plot and histo of calibration values
#define PLOT_WAVE 0

// Other info for calibration code
#define FIT_EN 1                //
#define FIT_WAVE_EN 1
#define OUTPUT_GAIN 1           // write full-run gains to file
#define OUTPUT_REPORT 1         // write full report including all fits and gains

// Definitions of types
typedef std::map < std::vector<int>, float > ReferenceValueMap;
typedef std::pair <std::vector<int>, float > ReferenceValuePair;
typedef std::map <std::vector<int>, float> :: iterator ReferenceValueMapIt;

// Define structure for storing full configuration 
struct RunConfig {              // this struct will hold all information 

   // Settings related to how the code works
   // ---------------------------------------
   // Which parts of code to run
   bool RunCalibration;
   bool RunSpecCal;
   bool RunSegCoreCorrelation;
   bool RunEfficiency;
   bool RunSpecEff;
   bool RunPropCrosstalk;
   bool RunWaveform;
   bool RunDiffCrosstalk;
   bool RunGeTiming;
   // Input files
    std::vector < std::string > files;
   // root output files
    std::string OutPath;
   // Information on energy and wave calibration files
    std::string EnergyCalibrationFile;
   bool HaveAltEnergyCalibration;
    std::string WaveCalibrationFile;
   bool HaveWaveCalibration;
   // What to print and how often
   bool PrintBasic;
   int PrintFrequency;
   bool PrintVerbose;
   // Limit on number of events
   int EventLimit;
   // ROOT stuff
   unsigned int ROOT_MaxVirtSize;
   // Configuration file
   std::string ConfigFile;

   // Global physics settings
   // ------------------------------------------
   // Properties of waveforms stored in the data
   unsigned int WaveformSamples;
   unsigned int WaveInitialSamples;
   unsigned int WaveFinalSamples;
   // integration in charge evaluation
   unsigned int Integration;
   // Dispersion used in charge evaluation
      // This is the dispersion stored in odb and used in charge evaluation/analyser spectra
      // not to be confused with dispersion in histogram calibration codes which is derived from max/bins
      // this one is needed to get the correct correspondance between wave charge and fpga/tig10 charge
   unsigned int Dispersion;
   // source Information
    std::vector < std::vector < float >>Sources;
    std::vector < std::string > SourceNames;
    std::vector < int >SourceNumCore;
    std::vector < int >SourceNumFront;
    std::vector < int >SourceNumBack;
   // Energy thresholds
   float EnergyThresh;
   int ChargeThresh;
   // Array mode
   bool HighEffMode;            // 1 if TIGRESS is in high eff mode (11cm), 0 if in high peak/total (14.5cm)
   
   // Calibration Stuff
   // Alternate Energy Calibration 
   vector < string > EnCalibNames;
   vector < vector < float >>EnCalibValues;
   // Waveform calibration
   vector < string > WaveCalibNames;
   vector < vector < float >>WaveCalibValues;

   // Settings for saving events to root file
   std::string EventOut;

   // Settings for individual functions
   // ------------------------------------------

   // SortTrees:Calib.C and SortHistos:HistCalib.C & SegCoreCalib.C 
   //----------------------------------------------------------------
   // What to do
   bool CalEnergy;              // fit charge spectra
   bool CalWave;                // fit wave-charge spectra
   bool Cal2D;                  // create 2D core-seg energy matrices for low stat seg calibration
   bool CalCheck2D;              // create 2D spectra for comparing wave charge and fpga evaluated charge
   bool CalReport;              // write full report on peak fitting as well as list of gains
   bool CalFile;                // produce a .cal file, readable by GRSISpoon
   bool CalList[CLOVERS][CRYSTALS][SEGS + 2];   // mask to determine which channels CalibSpectra() should run on
   bool CalListProvided;
   bool CalOutputBad;           // output to gain file if calibration is bad?
   bool CalOverwriteBad;           // Overwrite bad calibration values with mean (useful to get bad seg cals roughly right)
   // Output
    std::string CalOut;         // First 8 letters of this should match any file loaded in to offline cal
    std::string CalSpecOut;
    std::string CalName;
    std::string AnaName;
   bool WriteFits;              // --cal: write histo after fits, --calof: write new file with histos and fits
   // Gain drift/time dependent stuff
   float MaxTime;        // in seconds, max time after start of run.
   unsigned int TimeBins;
   float TimeBinSize;
   bool FitTempSpectra;  // should the temporary charge spectra be fitted for gain drift check
   // Charge spectra
   int ChargeBins;
   int ChargeBins2D;
   float ChargeMax;
   float WaveChargeMax;
   // Other spectra
   int FoldMax;
   // What to plot
   bool PlotFits;
   bool CalibPlots[CLOVERS][CRYSTALS][SEGS + 2];        // records if fits should be plotted for each channel
   bool PlotCalib;
   bool PlotResidual;
   bool PlotCalibSummary;
   // Fit Options
   // Peak Search stuff (passed to ROOT's TSpectrum->Search() method)
   float EnSearchThresh;
   float EnSearchSigma;
   float WaveSearchThresh;
   float WaveSearchSigma;
   // Clear low bins to aid peak search (large spike at zero can interfere
   bool ClearLowBinsWave;
   bool ClearLowBinsEnergy;
   int MaxClearBin;
   // Energy/ch fitting 
   // Estimate of gain for input of fit
   float TIGGainEst;
   float TIGWaveGainEst;
   // Extra calibration point at 0 ch = 0 keV
   // Peak Fitting
   unsigned int MinFitCounts;  // Minimum counts in whole spectrum for fit to be attempted
   float FitWidth_keV;           // Width of region either side of peak to be fitted
   bool FitBackground;           // 1 = yes, 0 = no.  Should be best to use this all the time but left option there just in case.
   float BackWidth_keV; 
   
   // Initial values for custom fit functions
   // These only effect the custom functions used if FIT_BACKGROUND == 1
   float EnergySigmaZero;
   float EnergySigma1MeV; 
   float WaveSigmaZero; 
   float WaveSigma1MeV;
   // Checks on fit quality
   float GausHeightMin;
   float GausCSPDMax;
   float GausSigmaMin;
   
   // Calibration options
   bool FitZero;                // Add extra calibration point at 0ch = 0keV
   float ZeroErr;
   bool ForceLinear;         // force linear calibration, even if numlines > 2
   bool ManualPeakSelect[CLOVERS][CRYSTALS][SEGS + 2];  // records if manual peak selection should be used
   bool ManualPeakCorrection;   // Manual peak selection if auto fails
   // Reference 
   ReferenceValueMap Crystal_FWHM_1332;  // # CRYSTAL_FHWM_1332
   
   // Additional options for SegCoreCalib.C
   // --------------------------------------
   int SegCoreFitOrder; // 0= gain only, 1=gain+offset, 2=gain+offset+quad
   bool PlotSegCoreCal;

   // CoincEff()
   //-----------
   // What to plot
   bool PlotEff;
   bool OutputEff;
   // Output
    std::string EffOut;
    std::string EffTxtOut;
   // Reference values
   ReferenceValueMap Sim_Clover_AB_Eff;  // # SIM_CLOVER_AB_EFF
   ReferenceValueMap Exp_Clover_AB_Eff;  // # EXP_CLOVER_AB_EFF
   ReferenceValueMap Sim_Crystal_Eff;    // # SIM_CRYSTAL_EFF
   ReferenceValueMap Exp_Crystal_Eff;    // # EXP_CRYSTAL_EFF
   
   bool EffHaveSimRef;
    std::string EffSimRefFileName;
   bool EffHaveExpRef;
    std::string EffExpRefFileName;

   // PropXtalk()
   //--------------
   // Output
    std::string PropOut;
    std::string PropTxtOut;
    
   // GeTiming()
   //-----------
   // Output
   std::string GeTimingOut;
   float GeTimingGateCentre;
   float GeTimingGateWidth;
};

// --------------------------------------------------------
// Variables:
// --------------------------------------------------------

extern RunConfig Config;


// --------------------------------------------------------
// Functions:
// --------------------------------------------------------
int LoadDefaultSettings();
int ReadCommandLineSettings(int argc, char **argv);
int ReadOptionsFile(std::string filename);
int ReadCloverRef(std::ifstream *File, ReferenceValueMap *Map);
int ReadCrystalRef(std::ifstream *File, ReferenceValueMap *Map);




